### PR TITLE
docker&singularity autobuild&autopublish

### PR DESCRIPTION
**Pull request by exaexa**

### CODE REVIEW

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 67799013 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 25a72aa0 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit e321d832 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit a78cf515 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 0f5c71db is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 30c2a01c is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 7e7b5fa3 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit c74b4b3a is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 57498efd is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit b39d6ca7 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 5e9d05e9 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 268bc048 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 07006419 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 2ece2ae4 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit d94eff63 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 60b31966 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit a8e22fd8 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit f76f5073 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 6c293dc6 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit c85856f4 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit a85be64c is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 3b83f968 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 7a4e5ee1 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit fe6f0533 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 42e48c55 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 8f2477ee is formatted properly.

**Comment by exaexa** [comment]
@laurentheirendt can you pls add cylon's token to secret `SYLABS_IO_TOKEN`?

(the token should be usable in `apptainer remote login` -- the command asks for one so you can test it)

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 6976af52 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 0469b45b is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 9db12fd5 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 8f968d78 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 5605e0de is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 320b68d3 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 9a082877 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 74e1ecb0 is formatted properly.

**Comment by exaexa** [comment]
OK so the thing actually builds and deploys as an artifact, should be reachable forever from gitlab if people want to download it automatically.

Ref: https://git-r3lab.uni.lu/lcsb-biocore/COBREXA.jl/-/pipelines/52220

The config now sets that the build happens:
- on `develop` branch, storing the `sif` for 1 week, so that we see if the container build is not b0rked before merging to master
- on `v*` tags

I'd normally squash the commit line but let's leave it here as a warning for others. Don't play with singu^H^H^H^H^Happtainer.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 1a26c294 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 7f2b4a6f is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit fdad7c0d is formatted properly.

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -3,21 +3,55 @@ name: Publish Docker image
...
+      - develop
+    tags:
+      - "v*"
 
 jobs:
   push_to_registry:
-    if: ${{ startsWith(github.ref, 'refs/tags/v') }}
     name: Push Docker image to GitHub Packages
-    runs-on: ubuntu-latest
+    runs-on: "ubuntu-20.04"
```
Comment (laurentheirendt) [inline .github/workflows/docker.yml:15]: I am wondering if we should not leave `latest` here

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -3,21 +3,55 @@ name: Publish Docker image
...
+        with:
+          username: ${{ secrets.DOCKER_IO_USER }}
+          password: ${{ secrets.DOCKER_IO_ACCESS_TOKEN }}
+
+      - name: Build and push to DockerHub
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: true
+          tags: "${{ secrets.DOCKER_IO_USER }}/cobrexa.jl:latest"
```
Comment (laurentheirendt) [inline .github/workflows/docker.yml:34]: can we somewhat make `cobrexa.jl` as a variable name?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -3,21 +3,55 @@ name: Publish Docker image
...
         with:
-          image: cobrexa.jl
-          tags: latest
           registry: ghcr.io
-          username: cylon-x
-          password: ${{ secrets.docker_pwd }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
```
Comment (laurentheirendt) [inline .github/workflows/docker.yml:44]: can you add here the link of the reference so that we don't forget?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -80,6 +80,16 @@ variables:
   script:
     - $ARTENOLIS_SOFT_PATH/julia/$JULIA_VER/Contents/Resources/julia/bin/julia --inline=yes --check-bounds=yes --color=yes --project=@. -e 'import Pkg; Pkg.test(; coverage = true)'
 
+.global_build_apptainer: &global_build_apptainer
+  stage: assets
+  image:
+    name: "quay.io/singularity/singularity:v3.9.4"
```
Comment (laurentheirendt) [inline .gitlab-ci.yml:86]: can you make the version a variable?

File: README.md [hunk 5]
```diff
@@ -5,12 +5,15 @@
 
 # COnstraint-Based Reconstruction and EXascale Analysis
 
-[docs-img-stable]: https://img.shields.io/badge/docs-stable-blue.svg
+[docs-img-stable]: https://img.shields.io/badge/docs-stable-86f.svg
```
Comment (laurentheirendt) [inline README.md:8]: I think the standard is `blue`

File: README.md [hunk 6]
```diff
@@ -23,9 +26,9 @@
 [repostatus-url]: https://www.repostatus.org/#active
 [repostatus-img]: https://www.repostatus.org/badges/latest/active.svg
 
-| **Documentation** | **Tests** | **Coverage** | **How to contribute?** | **Project status** |
-|:--------------:|:-------:|:---------:|:---------:|:---------:|
-| [![docs-img-stable]][docs-url-stable] [![docs-img-dev]][docs-url-dev] | [![CI][ci-img]][ci-url] | [![codecov][cov-img]][cov-url] | [![contrib][contrib-img]][contrib-url] | [![repostatus-img]][repostatus-url] |
+| **Documentation** | **Docker** | **Tests** | **Coverage** | **How to contribute?** | **Project status** |
+|:---:|:---:|:---:|:---:|:---:|:---:|
```
Comment (laurentheirendt) [inline README.md:30]: Can we also add the link to `apptainer`?

Comment (laurentheirendt) [reply README.md:30]: maybe separate list/table could do

File: cobrexa.def [hunk 7]
```diff
@@ -1,11 +1,11 @@
-Bootstrap: library
+BootStrap: library
```
Comment (laurentheirendt) [inline cobrexa.def:1]: Is this change important?

**Review by laurentheirendt — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (exaexa) [reply README.md:30]: Converted to a separate subsection

File: cobrexa.def (hunk 7 above)
Comment (exaexa) [reply cobrexa.def:1]: not sure, docs are using BootStrap and this remained here from trying to fix another issue. Will revert

File: .github/workflows/docker.yml (hunk 2 above)
Comment (exaexa) [reply .github/workflows/docker.yml]: done, we can actually use another metadata extraction action to make this much better (also with tags)

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit da5ec053 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit a14daf7f is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit c68f07ae is formatted properly.

**Review by laurentheirendt — APPROVED** [review]

**Pull request merged**